### PR TITLE
[SPARK-19439][PYSPARK][SQL] PySpark's registerJavaFunction Should Support UDAFs

### PR DESCRIPTION
**Pull request by zjffdu**
## What changes were proposed in this pull request?

Support register Java UDAFs in PySpark so that user can use Java UDAF in PySpark. Besides that I also add api in `UDFRegistration`

## How was this patch tested?

Unit test is added

### CODE REVIEW

**Comment by zjffdu** [comment]
@holdenk @marmbrus Please help review

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 1]
```diff
@@ -484,6 +484,21 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
 
   }
 
+  private[sql] def registerJavaUDAF(name: String, className: String): Unit = {
```
Comment (holdenk) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: It would be good to have ScalaDoc/JavaDoc for this method (similar to the current registerJava function) even though its not external facing, documenting it for whoever needs to update this next will make things easier.

Comment (zjffdu) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: ScalaDoc is added

**Comment by zjffdu** [comment]
@holdenk Mind to review it ?

File: python/pyspark/sql/context.py [hunk 2]
```diff
@@ -228,6 +228,24 @@ def registerJavaFunction(self, name, javaClassName, returnType=None):
...
+
+        :param name:  name of the UDF
+        :param javaClassName: fully qualified name of java class
+
+        >>> sqlContext.registerJavaUDAF("javaUDAF",
+        ...   "org.apache.spark.sql.hive.aggregate.MyDoubleAvg")
+        >>> df = sqlContext.createDataFrame([(1, "a"),(2, "b"), (3, "a")],["id", "name"])
+        >>> df.registerTempTable("df")
+        >>> sqlContext.sql("SELECT name,javaUDAF(id) as avg from df group by name").collect()
+        [Row(name=u'b', avg=102.0), Row(name=u'a', avg=102.0)]
```
Comment (viirya) [inline python/pyspark/sql/context.py]: Shall we add a test failing the registration? E.g. an incorrect class name.

File: python/pyspark/sql/context.py [hunk 3]
```diff
@@ -228,6 +228,24 @@ def registerJavaFunction(self, name, javaClassName, returnType=None):
...
+        """Register a java UDAF so it can be used in SQL statements.
+
+        :param name:  name of the UDF
+        :param javaClassName: fully qualified name of java class
+
+        >>> sqlContext.registerJavaUDAF("javaUDAF",
+        ...   "org.apache.spark.sql.hive.aggregate.MyDoubleAvg")
+        >>> df = sqlContext.createDataFrame([(1, "a"),(2, "b"), (3, "a")],["id", "name"])
+        >>> df.registerTempTable("df")
+        >>> sqlContext.sql("SELECT name,javaUDAF(id) as avg from df group by name").collect()
```
Comment (viirya) [inline python/pyspark/sql/context.py]: nit: name,javaUDAF(id) -> name, javaUDAF(id)

File: python/pyspark/sql/context.py [hunk 4]
```diff
@@ -228,6 +228,24 @@ def registerJavaFunction(self, name, javaClassName, returnType=None):
             jdt = self.sparkSession._jsparkSession.parseDataType(returnType.json())
         self.sparkSession._jsparkSession.udf().registerJava(name, javaClassName, jdt)
 
+    @ignore_unicode_prefix
+    @since(2.2)
+    def registerJavaUDAF(self, name, javaClassName):
+        """Register a java UDAF so it can be used in SQL statements.
+
+        :param name:  name of the UDF
```
Comment (viirya) [inline python/pyspark/sql/context.py]: name of the UDAF.

File: python/pyspark/sql/context.py (hunk 2 above)
Comment (zjffdu) [inline python/pyspark/sql/context.py]: Good point, will add that

File: python/pyspark/sql/tests.py [hunk 5]
```diff
@@ -436,6 +436,20 @@ def test_udf_with_order_by_and_limit(self):
...
+    def test_non_existed_udf(self):
+        try:
+            self.spark.udf.registerJavaFunction("udf1", "non_existed_udf")
+            self.fail("should fail due to can not load java udf class")
+        except py4j.protocol.Py4JError as e:
+            self.assertTrue("Can not load class non_existed_udf" in e.desc)
+
+    def test_non_existed_udaf(self):
+        try:
+            self.spark.udf.registerJavaFunction("udf1", "non_existed_udaf")
```
Comment (viirya) [inline python/pyspark/sql/tests.py]: registerJavaUDAF?

Comment (zjffdu) [inline python/pyspark/sql/tests.py]: Correct, fixed 😄

File: python/pyspark/sql/tests.py [hunk 6]
```diff
@@ -436,6 +436,20 @@ def test_udf_with_order_by_and_limit(self):
...
+    def test_non_existed_udf(self):
+        try:
+            self.spark.udf.registerJavaFunction("udf1", "non_existed_udf")
+            self.fail("should fail due to can not load java udf class")
+        except py4j.protocol.Py4JError as e:
+            self.assertTrue("Can not load class non_existed_udf" in str(e))
+
+    def test_non_existed_udaf(self):
+        try:
+            self.spark.udf.registerJavaUDAF("udf1", "non_existed_udaf")
```
Comment (viirya) [inline python/pyspark/sql/tests.py]: nit: udf1 -> udaf1.

**Comment by viirya** [comment]
LGTM, see if @marmbrus or @holdenk have any more comments about this change.

**Comment by zjffdu** [comment]
@viirya Thanks for careful review.

**Comment by holdenk** [comment]
I'll try and follow up this weekend.

**Comment by viirya** [comment]
retest this please.

**Comment by holdenk** [comment]
Thanks for working on this, one thing is the return's from the `UDFRegistration` functions seem like they won't do much since the functions they call on the sqlContext don't have returns (and they probably should for usability) - unless I'm missing something.

**Comment by zjffdu** [comment]
Good catch ! @holdenk `return` is removed.

**Comment by zjffdu** [comment]
ping @holdenk

**Comment by holdenk** [comment]
@zjffdu - how would you feel about putting the return back, and just plumbing it through as required? It seems like it would be useful to have users able to programmatically do this (I find my self effectively doing this in some of my own personnel notebooks)?

**Comment by zjffdu** [comment]
@holdenk But it has nothing to return, because scala side return Unit.  See https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala#L528

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 7]
```diff
@@ -475,20 +475,42 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
...
+      case e: ClassNotFoundException => throw new IOException(s"Can not load class ${className}, please make sure it is on the classpath")
     }
 
   }
 
   /**
+   * Register a Java UDAF class using reflection, for use from pyspark
+   *
+   * @param name     UDAF name
+   * @param className    fully qualified class name of UDAF
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: Missing @since.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 8]
```diff
@@ -475,20 +475,42 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
...
             case 22 => register(name, udf.asInstanceOf[UDF21[_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _]], returnType)
             case 23 => register(name, udf.asInstanceOf[UDF22[_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _]], returnType)
-            case n => logError(s"UDF class with ${n} type arguments is not supported ")
+            case n =>
+              throw new IOException(s"UDF class with ${n} type arguments is not supported.")
           }
         } catch {
           case e @ (_: InstantiationException | _: IllegalArgumentException) =>
-            logError(s"Can not instantiate class ${className}, please make sure it has public non argument constructor")
+            throw new IOException(s"Can not instantiate class ${className}, please make sure it has public non argument constructor")
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: Please throw an `AnalysisException`

**Comment by gatorsmile** [comment]
Will review this PR more carefully in the next few days.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala (hunk 7 above)
Comment (zjffdu) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: @since is needed for private function ?

**Comment by holdenk** [comment]
@zjffdu - if you look at https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala#L129  though it returns the `UserDefinedFunction` (currently the Python one is Unit but it would be more useful if it returned a `UserDefinedFunction`). I think to make it easier for people to take advantage of Java UDFs we would want them to be able to use it programmatic ally in the Dataframe DSL not just in SQL string expressions.

What do you think @gatorsmile & @zjffdu ?

**Comment by zjffdu** [comment]
@holdenk The link you pasted is for the case that using scala closure to create udf. While `registerJava` use java reflection to create udf. This is what I use in `registerJava` https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala#L528 It returns Unit.
Maybe it is possible to create `registerScala` to return scala udf. But it seems it is not possible for java udf.

**Comment by zjffdu** [comment]
@holdenk @gatorsmile Any more comments ?

**Comment by cloud-fan** [comment]
Hi @zjffdu thanks for working on it! But I'm not sure how useful this feature will be. AFAIK most users use scala/java UDF instead of Python UDF because it's too slow. We are working on a project to improve the communication between JVM and Python process, which may add a new Python UDF interface and also affect the python UDAF design. Can you hold this PR for a while? Thanks!

**Comment by zjffdu** [comment]
@cloud-fan This is not about using python UDF, it is to allow pyspark to use java UDF (no python daemon will be launched). So actually it would improve the performance.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 9]
```diff
@@ -475,20 +475,42 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
...
 
   }
 
   /**
+   * Register a Java UDAF class using reflection, for use from pyspark
+   *
+   * @param name     UDAF name
+   * @param className    fully qualified class name of UDAF
+   */
+  private[sql] def registerJavaUDAF(name: String, className: String): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: do we need `returnType`?

Comment (zjffdu) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: pyspark side don't need `returnType` so I didn't use `returnType` here, and it is private function so should be open for adding `returnType` in future if necessary.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: at python side, why do we need returnType in `registerJavaFunction`?

Comment (zjffdu) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: I mean `registerJavaUDAF` in `context.py` does't have `returnType`, so here in scala side, I don't provide `returnType `either since this scala method is only used for `registerJavaUDAF` of pyspark

Comment (zjffdu) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: This due to in scala side `registerJava` of `UDFRegistration' needs returnType. Yeah, it do looks like a little weird for python side to require returnType.

**Comment by gatorsmile** [comment]
How about creating a new suite `JavaUDAFSuite.java`, like `JavaUDFSuite.java`?

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala (hunk 9 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: `registerJava` accepts an optional return type, if not given, spark will try to infer it via reflection. do we really not need to do this for udaf?

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: `UserDefinedAggregateFunction` already defines its return type. So we don't need to specify it when registering a udfa.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 10]
```diff
@@ -491,20 +491,42 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
             case 21 => register(name, udf.asInstanceOf[UDF20[_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _]], returnType)
             case 22 => register(name, udf.asInstanceOf[UDF21[_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _]], returnType)
             case 23 => register(name, udf.asInstanceOf[UDF22[_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _]], returnType)
-            case n => logError(s"UDF class with ${n} type arguments is not supported ")
+            case n =>
+              throw new IOException(s"UDF class with ${n} type arguments is not supported.")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: why IOException?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala [hunk 11]
```diff
@@ -20,16 +20,19 @@ package org.apache.spark.sql.hive.execution
 import scala.collection.JavaConverters._
 import scala.util.Random
 
+import _root_.test.org.apache.spark.sql.MyDoubleAvg
+import _root_.test.org.apache.spark.sql.MyDoubleSum
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: `_root_`?

Comment (zjffdu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: oops, fixed.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala (hunk 10 above)
Comment (zjffdu) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: I didn't find a more proper exception type, so just use IOException.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: Please use `AnalysisException`, as suggested above.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala (hunk 7 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: OK. I did not notice it.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala (hunk 10 above)
Comment (zjffdu) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala]: Sorry, miss your last comment, fixed

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala [hunk 12]
```diff
@@ -20,16 +20,19 @@ package org.apache.spark.sql.hive.execution
 import scala.collection.JavaConverters._
 import scala.util.Random
 
+import test.org.apache.spark.sql.MyDoubleAvg
+import test.org.apache.spark.sql.MyDoubleSum
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: Revert the changes made in this file?

File: python/pyspark/sql/context.py [hunk 13]
```diff
@@ -232,6 +232,23 @@ def registerJavaFunction(self, name, javaClassName, returnType=None):
             jdt = self.sparkSession._jsparkSession.parseDataType(returnType.json())
         self.sparkSession._jsparkSession.udf().registerJava(name, javaClassName, jdt)
 
+    @ignore_unicode_prefix
+    @since(2.2)
```
Comment (gatorsmile) [inline python/pyspark/sql/context.py]: `2.3`?

**Comment by gatorsmile** [comment]
Any update on this? Thanks!

**Comment by zjffdu** [comment]
@gatorsmile sorry for late response, will update it soon

**Comment by holdenk** [comment]
Just a gentle ping, I've got some review bandwidth coming up and I'd love to see us get this in :)

**Comment by gatorsmile** [comment]
cc @ueshin

File: python/pyspark/sql/tests.py [hunk 14]
```diff
@@ -481,6 +481,20 @@ def test_udf_registration_returns_udf(self):
             df.select(add_three("id").alias("plus_three")).collect()
         )
 
+    def test_non_existed_udf(self):
+        try:
+            self.spark.udf.registerJavaFunction("udf1", "non_existed_udf")
+            self.fail("should fail due to can not load java udf class")
+        except AnalysisException as e:
```
Comment (cloud-fan) [inline python/pyspark/sql/tests.py]: shall we use `self.assertRaises` like other tests?

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameSuite.java [hunk 15]
```diff
@@ -40,6 +40,8 @@
 import org.apache.spark.sql.types.*;
 import org.apache.spark.util.sketch.BloomFilter;
 import org.apache.spark.util.sketch.CountMinSketch;
+import test.org.apache.spark.sql.MyDoubleAvg;
+import test.org.apache.spark.sql.MyDoubleSum;
```
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameSuite.java]: unnecessary change?

File: sql/core/src/test/java/test/org/apache/spark/sql/MyDoubleAvg.java [hunk 16]
```diff
@@ -28,6 +25,9 @@
 import org.apache.spark.sql.types.StructField;
 import org.apache.spark.sql.types.StructType;
 
+import java.util.ArrayList;
+import java.util.List;
```
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/MyDoubleAvg.java]: the import order is wrong here, please follow the previous style

File: sql/core/src/test/java/test/org/apache/spark/sql/MyDoubleSum.java [hunk 17]
```diff
@@ -15,18 +15,18 @@
...
-import org.apache.spark.sql.types.StructField;
-import org.apache.spark.sql.types.StructType;
 import org.apache.spark.sql.types.DataType;
 import org.apache.spark.sql.types.DataTypes;
-import org.apache.spark.sql.Row;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+import java.util.ArrayList;
+import java.util.List;
```
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/MyDoubleSum.java]: ditoo

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala [hunk 18]
```diff
@@ -20,16 +20,19 @@ package org.apache.spark.sql.hive.execution
...
 import org.apache.spark.sql.catalyst.expressions.UnsafeRow
 import org.apache.spark.sql.expressions.{MutableAggregationBuffer, UserDefinedAggregateFunction}
 import org.apache.spark.sql.functions._
-import org.apache.spark.sql.hive.aggregate.{MyDoubleAvg, MyDoubleSum}
 import org.apache.spark.sql.hive.test.TestHiveSingleton
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.SQLTestUtils
 import org.apache.spark.sql.types._
 
+
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: shall we move this test suite to sql/core?

File: sql/hive/src/test/java/org/apache/spark/sql/hive/JavaDataFrameSuite.java [hunk 19]
```diff
@@ -31,7 +31,7 @@
 import org.apache.spark.sql.expressions.UserDefinedAggregateFunction;
 import static org.apache.spark.sql.functions.*;
 import org.apache.spark.sql.hive.test.TestHive$;
-import org.apache.spark.sql.hive.aggregate.MyDoubleSum;
+import test.org.apache.spark.sql.MyDoubleSum;
 
 public class JavaDataFrameSuite {
```
Comment (cloud-fan) [inline sql/hive/src/test/java/org/apache/spark/sql/hive/JavaDataFrameSuite.java]: shall we merge this suite with `JavaDataFrameSuite` in sql/core?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala (hunk 18 above)
Comment (zjffdu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: I didn't add any test in this file. Or do you mean move AggregationQuerySuite.scala to sql/core ?

File: sql/hive/src/test/java/org/apache/spark/sql/hive/JavaDataFrameSuite.java (hunk 19 above)
Comment (zjffdu) [inline sql/hive/src/test/java/org/apache/spark/sql/hive/JavaDataFrameSuite.java]: do you mean move JavaDataFrameSuite to sql/core ?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala (hunk 18 above)
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: yea, seems there is no reason to leave this suite in sql/hive

File: sql/hive/src/test/java/org/apache/spark/sql/hive/JavaDataFrameSuite.java (hunk 19 above)
Comment (cloud-fan) [inline sql/hive/src/test/java/org/apache/spark/sql/hive/JavaDataFrameSuite.java]: yea

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala (hunk 18 above)
Comment (zjffdu) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: It depends on some hive stuff (`TestHiveSingleton`), so I guess it is intended to be put in sql/hive.

Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: when move it to sql/core, we can make it extend `SharedSQLContext`.

**Comment by zjffdu** [comment]
This PR fails fails PySpark pip packaging tests. But I don't know what's wrong here. @holdenk Is the `PySpark pip packaging test` an known issue ?

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
@zjffdu Please refer to the comments of #15821, especially the comments after https://github.com/apache/spark/pull/15821#issuecomment-310894657.

**Comment by zjffdu** [comment]
Thanks @viirya

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by dongjoon-hyun** [comment]
Hi, @zjffdu .

After this commit, Maven shows the following warning messages due to the change of `sql/hive/pom.xml`. Could you check that?

```bash
$ build/mvn -Phadoop-2.6 -Pmesos -Pkinesis-asl -Pyarn -Phive-thriftserver -Phive clean package -DskipTests
Using `mvn` from path: /usr/local/bin/mvn
[INFO] Scanning for projects...
[WARNING]
[WARNING] Some problems were encountered while building the effective model for org.apache.spark:spark-hive_2.11:jar:2.3.0-SNAPSHOT
[WARNING] 'dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: org.apache.spark:spark-sql_${scala.binary.version}:test-jar -> duplicate declaration of version ${project.version} @ org.apache.spark:spark-hive_2.11:[unknown-version], /Users/dongjoon/SPARK-POM/sql/hive/pom.xml, line 177, column 17
[WARNING]
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING]
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
...
```
